pypdf/_text_extraction/_layout_mode/_fixed_width_page.py

```diff
@@ -0,0 +1,398 @@
+        elif op == b"'":
+            xform_stack.reset_trm()
+            xform_stack.add_tm([0, xform_stack.TL])
+            tj_ops.append(decode_tj(opands[0], xform_stack))
+        elif op == b'"':
+            xform_stack.reset_trm()
+            _set_state_param(b"Tw", [opands[0]], xform_stack)
+            _set_state_param(b"Tc", [opands[1]], xform_stack)
+            xform_stack.add_tm([0, xform_stack.TL])
+            tj_ops.append(decode_tj(opands[2], xform_stack))
+        elif op in (b"Td", b"Tm", b"TD", b"T*"):
+            xform_stack.reset_trm()
+            if op == b"Tm":
+                xform_stack.reset_tm()
+            elif op == b"TD":
+                _set_state_param(b"TL", [-opands[1]], xform_stack)
+            elif op == b"T*":
+                opands = [0, -xform_stack.TL]
+            xform_stack.add_tm(opands)
+        elif op == b"Tf":
+            xform_stack.font_size = opands[1]
+            xform_stack.font = fonts[opands[0]]
+        else:
+            _set_state_param(op, opands, xform_stack)
+
+
+def y_coordinate_groups(
+    bt_groups: List[BTGroup], debug_path: Union[Path, None] = None
+) -> Dict[int, List[BTGroup]]:
+    """
+    Group text operations by rendered y coordinate, i.e. the line number
+
+    Args:
+        bt_groups: list of dicts as returned by text_show_operations()
+        debug_file: full path + filename prefix for debug output. Defaults to None.
+
+    Returns:
+        Dict[int, List[BTGroup]]: dict of lists of text rendered by each BT operator
+            keyed by y coordinate
+    """
+    ty_groups = {
+        ty: sorted(grp, key=lambda x: x["tx"])
+        for ty, grp in groupby(
+            bt_groups, key=lambda bt_grp: int(bt_grp["ty"] * bt_grp["flip_sort"])
+        )
+    }
+    # combine groups whose y coordinates differ by less than the effective font height
+    # (accounts for mixed fonts and other minor oddities)
+    last_ty = next(iter(ty_groups))
+    last_txs = {int(_t["tx"]) for _t in ty_groups[last_ty] if _t["text"].strip()}
+    for ty in list(ty_groups)[1:]:
+        fsz = min(ty_groups[_y][0]["font_height"] for _y in (ty, last_ty))
+        txs = {int(_t["tx"]) for _t in ty_groups[ty] if _t["text"].strip()}
+        # prevent merge if both groups are rendering in the same x position.
+        no_text_overlap = not (txs & last_txs)
+        offset_less_than_font_height = abs(ty - last_ty) < fsz
+        if no_text_overlap and offset_less_than_font_height:
+            ty_groups[last_ty] = sorted(
+                ty_groups.pop(ty) + ty_groups[last_ty], key=lambda x: x["tx"]
+            )
+            last_txs |= txs
+        else:
+            last_ty = ty
+            last_txs = txs
+    if debug_path:
+        debug_path.with_name("bt_groups.json").write_text(
+            json.dumps(ty_groups, indent=2, default=str), "utf-8"
+        )
+    return ty_groups
+
+
+def _set_state_param(op: bytes, opands: List[Any], xform_stack: XformStack) -> None:
```

Comment:
Why `opands` instead of `operands`?

Reply:
In an early version, use of "operands" caused `redefined-outer-name` errors, so I started using opands. I don't feel like it detracts from readability / understandability and it's less typing but happy to do a search / replace if desired. Shouldn't hurt anything at this point.

Reply:
Universally renamed as `operands` for clarity.